src/gui/qgsmaptooldigitizefeature.h

```diff
@@ -17,31 +17,65 @@
 #define QGSMAPTOOLDIGITIZEFEATURE_H
 
 #include "qgsmaptoolcapture.h"
-#include "qgis_app.h"
+#include "qgis_gui.h"
 
 class QgsFeature;
 
-//! This tool digitizes geometry of new point/line/polygon features on already existing vector layers
-class APP_EXPORT QgsMapToolDigitizeFeature : public QgsMapToolCapture
+/**
+ * \ingroup gui
+ * \brief This tool digitizes geometry of new point/line/polygon features on already existing vector layers
+ * Once the map tool is enable, user can digitize the feature geometry
+ * A signal will then be emitted.
+ * \since QGIS 3.10
+ */
+class GUI_EXPORT QgsMapToolDigitizeFeature : public QgsMapToolCapture
 {
     Q_OBJECT
+
   public:
-    //! \since QGIS 3.2
-    QgsMapToolDigitizeFeature( QgsMapCanvas *canvas, QgsMapLayer *layer, CaptureMode mode );
+
+    /**
+     * \brief QgsMapToolDigitizeFeature is a map tool to digitize a feature geometry
+     * \param canvas the map canvas
+     * \param cadDockWidget widget to setup advanced digitizing parameters
+     * \param mode type of geometry to capture (point/line/polygon), QgsMapToolCapture::CaptureNone to autodetect geometry
+     */
+    QgsMapToolDigitizeFeature( QgsMapCanvas *canvas, QgsAdvancedDigitizingDockWidget *cadDockWidget, CaptureMode mode = QgsMapToolCapture::CaptureNone );
 
     void cadCanvasReleaseEvent( QgsMapMouseEvent *e ) override;
 
-    virtual void digitized( QgsFeature &f );
+    /**
+     * Change the layer edited by the map tool
+     * \param vl the layer to be edited by the map tool
+     * \since 3.10
+     */
+    void setLayer( QgsMapLayer *vl ) { mLayer = vl; }
 
     void activate() override;
     void deactivate() override;
 
   signals:
-    void digitizingCompleted( const QgsFeature & );
+
+    /**
+     * Emitted whenever the digitizing has been successfully completed
+     * \param feature the new digitized feature
+     */
+    void digitizingCompleted( const QgsFeature &feature );
+
+    /**
+     * Emitted whenever the digitizing has been ended without digitizing
+     * any feature
+     */
     void digitizingFinished( );
 
   protected:
 
+    /**
+     * Finish the feature digitizing
+     * \param f the new created feature
+     */
+    virtual void digitized( QgsFeature &f );
```

Comment:
private?

Reply:
`const QgsFeature`?

Reply:
> private?

The digitized method is inherited by the [QgsMapToolAddFeature](https://github.com/troopa81/QGIS/blob/feature_createfeat_from_relationrefwidget/src/app/qgsmaptooladdfeature.cpp#L62)

> const QgsFeature?

Could be. But we will need to copy it later in order to modify it (when we display the form for instance). I didn't modify the original code.

Reply:
used by or implemented by?
Private virtual can be overridden.

The question regarding const is if the implementing method is expected to change the data. If yes, the meaning and effect of such a change should be documented.

Reply:
Done